include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2023 Cirrus Logic, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CHARGER_H_
+
+/**
+ * @brief Charger Interface
+ * @defgroup charger_interface Charger Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+/**
+ * @brief Runtime Dynamic Battery Parameters
+ */
+enum charger_props {
+	/** Indicates if external supply is present for the charger. */
+	/** Value should be of type enum charger_online */
+	CHARGER_PROP_ONLINE = 0,
+	/** Reports whether or not a battery is present. */
+	/** Value should be of type bool*/
+	CHARGER_PROP_PRESENT,
+	/** Represents the charging status of the charger. */
+	/** Value should be of type enum charger_status */
+	CHARGER_PROP_STATUS,
+	/** Reserved to demark end of common charger properties */
+	CHARGER_PROP_COMMON_COUNT,
+	/**
+	 * Reserved to demark downstream custom properties - use this value as the actual value may
+	 * change over future versions of this API
+	 */
+	CHARGER_PROP_CUSTOM_BEGIN = CHARGER_PROP_COMMON_COUNT + 1,
+	/** Reserved to demark end of valid enum properties */
+	CHARGER_PROP_MAX = UINT16_MAX,
+};
+
+/**
+ * @brief External supply states
+ */
+enum charger_online {
+	/** External supply not presented */
+	CHARGER_ONLINE_OFFLINE = 0,
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Representation of a charger property value
+ */
+union charger_value {
+	enum charger_online online;
+	bool present;
+	enum charger_status status;
+	int custom_int_val;
+};
```

Comment:
I still think this is a fragile approach we should avoid. Instead, a variant-like solution would make things much better. Here's an untested code snippet on what I had in mind:

```c
/* available types */
enum charger_type {
    CHARGER_TYPE_BOOL,
    CHARGER_TYPE_UINT,
    ....
};

/* type/data container */
struct charger_value {
    enum charger_type type;
    union {
        /* for boolean flags */
        bool b;
        /* for unsigned values */
        unsigned int uint;
        ...
    } value;
};

/* some static initializer helpers, if needed ? */
#define CHARGER_VALUE_INIT_BOOL(_value) { .type = CHARGER_TYPE_BOOL, .value = { .b = (_value) }}
#define CHARGER_VALUE_INIT_UINT(_value) { .type = CHARGER_TYPE_UINT, .value = { .uint = (_value) }}

/* runtime type assertion helper */
#define CHARGER_ASSERT_TYPE(_v, _type) __ASSERT_NO_MSG((_v)->type == (_type))

/* get/set/reset helpers, for each type */

/* get boolean value (asserts if value does not contain a boolean) -> used by API clients */
static inline bool charger_value_get_bool(const struct charger_value *v)
{
    CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_BOOL);
    return v->value.b;
};

/* set boolean value (asserts if value is not a boolean type -> used by ? */
static inline void charger_value_set_bool(struct charger_value *v, bool val)
{
    CHARGER_ASSERT_TYPE(v, CHARGER_TYPE_BOOL);
    v->value.b = val;
}

/* set boolean value and type, ignores current value type -> used by drivers to set the value of a given value pointer */
static inline void charger_value_reset_bool(struct charger_value *v, bool val)
{
    v->type = CHARGER_TYPE_BOOL;
    v->value.b = val;
}

/* ... same get/set/reset for unsigned int would go here */
```

I think this would make API interface much cleaner. Each `CHARGER_PROP_*` should also document its associated `CHARGER_TYPE_*`.

Reply:
so were reverting back to the struct again and also needing maintain and rely on helpers to get/set/reset for every data type (bool, int, enum(s)). That's a pretty major rewrite. I don't know if I can make that happen right away, it really can't be a subsequent PR?

Reply:
This new struct is not the same that was initially proposed. It's a _variant type_ container. The API will still take prop + value container as arguments. This also standardizes how drivers write charger values, and how clients consume property values. In the future, if API gains a `set` operation, things will also be standardized.

Reply:
I understand it is different, but it adds complexity that we sought to remove. I'm working on it.

I guess preemptively, do you have issues with uint being an int? These are batteries we're talking about so current values are often signed and I would prefer to use a signed int rather than an unsigned int and a flag.

Reply:
Integer is fine, snippet is just an idea. Another topic are units (handled poorly in sensors by associating each prop to a unit+scale). For example, if passing a current, what should it be, A, mA, uA....?

Maybe this can serve as a good reference as well:
https://github.com/Linaro/step/blob/main/include/step/measurement/unit.h
https://github.com/Linaro/step/blob/main/include/step/measurement/measurement.h

cc: @microbuilder

Reply:
https://docs.kernel.org/power/power_supply_class.html
For ease of talking up to power_supply I am looking to follow the units specified here. I see no value in changing to mA and A would just not be useful.

Reply:
That's fine, if micro range works for all types of these device classes there's no need to overcomplicate the design.